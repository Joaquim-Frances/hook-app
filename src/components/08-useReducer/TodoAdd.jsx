import React from 'react'
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({handleAddTodo}) => {

    const [{ tarea }, handleInputChange, reset] = useForm({
        tarea:'',
    });

    const handleSubmit = (e) => {
        e.preventDefault();
        if(tarea.trim().length <= 1){
            return;
        }
        const newTodo = {
            id: new Date().getTime(),
            tarea: tarea,
            done: false
        };
        handleAddTodo( newTodo );
        reset();
    }

    return (
        <>
        <h3>Agregar Tarea</h3>
        <hr/>
        <form onSubmit={ handleSubmit }>
            <input onChange={handleInputChange} type="text" value={tarea} name="tarea" placeholder="Escribe la tarea..." autoComplete="off" className="form-control"/>
            <button type='submit' className="btn btn-outline-primary mt-1 btn-block"> AGREGAR </button>
        </form>   
        </>
    )
}
