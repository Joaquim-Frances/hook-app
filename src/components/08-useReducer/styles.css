body{
    padding: 50px;
}

p{
    cursor: pointer;
    margin: 0px;
}

.list-group-item{
    display: flex;
    flex-direction: row;
    justify-content: space-between;
}

.complete{
    text-decoration: line-through;
}