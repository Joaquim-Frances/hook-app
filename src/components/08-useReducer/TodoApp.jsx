import React, { useReducer } from 'react';
import { todoReducer } from './todoReducer';

//styles
import './styles.css';

const initialState = [{
    id: new Date().getTime(),
    tarea: 'Aprender React',
    done: false
}];

export const TodoApp = () => {

    const [todos, dispatch ] = useReducer(todoReducer, initialState)

    
    

    const handleSubmit = (e) => {
        e.preventDefault();
        
        const newTodo = {
            id: new Date().getTime(),
            tarea: 'Nueva Tarea',
            done: false
        };

        const action = {
            type: 'add',
            payload: newTodo,
        }

        dispatch( action );
    }

    return (
        <div>
            <h1>Todo List ( {todos.length} )</h1>
            <hr/>

           <div className="row">
                <div className="col-7">
                    <ul className='list-group list-group-flush'>
                    {
                        todos.map( (todo, i) => {
                            return(
                            <li key={todo.id} className='list-group-item'>
                                <p className='text-center complete'>{i+1}. {todo.tarea}</p> 
                                <button className='btn btn-danger'> Borrar </button>
                            </li>)
                        })
                    }
                    </ul>
                </div>
                <div className="col-5">
                    <h3>Agregar Tarea</h3>
                    <hr/>

                    <form onSubmit={ handleSubmit }>
                        <input type="text" name="tarea" placeholder="Escribe la tarea..." autoComplete="off" className="form-control"/>
                        <button type='submit' className="btn btn-outline-primary mt-1 btn-block"> AGREGAR </button>
                    </form>
                </div>

           </div>
        </div>
    )
}
