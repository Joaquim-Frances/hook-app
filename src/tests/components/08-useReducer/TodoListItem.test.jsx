import { shallow } from "enzyme"
import { TodoListItem } from "../../../components/08-useReducer/TodoListItem"
//import { handleToggle } from "../../../components/08-useReducer/TodoApp";

const handleDelete = jest.fn()
const handleToggle = jest.fn()

const todo = {
    id:4,
    tarea: 'Comprar usb',
    done: true,
}

const i = 0;



describe('Testing in TodoListItem component', () => {
    
    const wrapper = shallow( <TodoListItem todo={todo} i={i} handleDelete={handleDelete} handleToggle={handleToggle}/>);

    test('should display correclty', () => {

        expect(wrapper).toMatchSnapshot();
        
    })

    test('should call delete method', () => {

        wrapper.find('button').simulate('click');
        expect(handleDelete).toBeCalled()
        expect(handleDelete).toBeCalledWith(todo.id)
        
    })

    test('should call toggle method', () => {

        wrapper.find('p').simulate('click');
        expect(handleToggle).toBeCalled()
        expect(handleToggle).toBeCalledWith(todo.id)
        
    })

    test('should writte correctly the <p></p> text', () => {

        
        expect(wrapper.find('p').text().trim()).toBe(`${i+1}. ${todo.tarea}`)
        
    })

    test('should have "complete" className if done is true', () => {

        wrapper.find('p').simulate('click');
        expect(wrapper.find('p').prop('className')).toBe('complete')
        expect(wrapper.find('p').hasClass('complete')).toBe(true)
        
    })
    
    
    
    
    
})
