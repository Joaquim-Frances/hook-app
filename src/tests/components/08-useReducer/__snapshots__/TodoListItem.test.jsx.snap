// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing in TodoListItem component should display correclty 1`] = `
<li
  className="list-group-item"
  key="4"
>
  <p
    className="complete"
    onClick={[Function]}
  >
    1
    . 
    Comprar usb
  </p>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
     Borrar 
  </button>
</li>
`;
